ownCloud/Client/Actions/Actions+Extensions/UploadPhotosAction.swift

```diff
@@ -95,59 +121,82 @@ class UploadPhotosAction: UploadBaseAction {
-					let storeAsHEIF = ressource.uniformTypeIdentifier == "public.heic" && !userDefaults.convertHeic
-					let colorSpace = CGColorSpaceCreateDeviceRGB()
+		// Setup import options, allow download asset from network if necessary
+		let contentInputOptions = PHContentEditingInputRequestOptions()
+		contentInputOptions.isNetworkAccessAllowed = true
+		contentInputOptions.progressHandler = { (percentage:Double, _) in
+			progress.completedUnitCount = Int64(percentage * 100)
+		}
 
-					var imageData : Data?
-					if storeAsHEIF {
-						imageData = CIContext().heifRepresentation(of: fullImage, format: CIFormat.RGBA8, colorSpace: colorSpace)
-						print(fullImage.properties)
-					} else {
-						localURL = localURL.deletingPathExtension().appendingPathExtension("jpg")
-						imageData = CIContext().jpegRepresentation(of: fullImage, colorSpace: colorSpace)
-					}
-					do {
-						self.unpublish(progress: progress)
-						try imageData?.write(to: localURL)
-						self.upload(itemURL: localURL, to: rootItem, name: localURL.lastPathComponent, completionHandler: { (_) in
-							// Delete the temporary asset file
-							try? FileManager.default.removeItem(at: localURL)
-							self.completed()
-						})
-					} catch {
-						self.completed(with: NSError(ocError: .internal))
+		// Import full size asset
+		asset.requestContentEditingInput(with: contentInputOptions) { (input, _) in
+			self.unpublish(progress: progress)
+
+			self.uploadSerialQueue.async {
+				if let input = input, let url = input.fullSizeImageURL {
+					autoreleasepool {
+						var image = CIImage(contentsOf: url)
+						if image != nil {
+							let storeAsHEIF = input.uniformTypeIdentifier == "public.heic" && !userDefaults.convertHeic
+							let colorSpace = CGColorSpaceCreateDeviceRGB()
+
+							let fileName = url.lastPathComponent
+							var localURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(fileName)
+
+							var ciContext = CIContext()
+							var imageData : Data?
+
+							func cleanUpCoreImageRessources() {
+								// Release memory consuming resources
+								imageData = nil
+								image = nil
+								ciContext.clearCaches()
+							}
+
+							if storeAsHEIF {
+								imageData = ciContext.heifRepresentation(of: image!, format: CIFormat.RGBA8, colorSpace: colorSpace)
+							} else {
+								localURL = localURL.deletingPathExtension().appendingPathExtension("jpg")
+								imageData = ciContext.jpegRepresentation(of: image!, colorSpace: colorSpace)
+							}
+							do {
+								// First write an image to a file stored in temporary directory
+								try imageData?.write(to: localURL)
+
+								cleanUpCoreImageRessources()
+
+								// Upload to the cloud
+								self.upload(itemURL: localURL, to: rootItem, name: localURL.lastPathComponent, completionHandler: { (actionPerformed) in
+									// Delete the temporary asset file
+									try? FileManager.default.removeItem(at: localURL)
+									completion(actionPerformed)
```

Comment:
The `completionHandler` will only be called by the upload-method after the image has finished uploading, so you shouldn't rely on it.

To catch critical errors (like I/O errors moving the file to local storage), please use the `placeholderCompletionHandler`. Right now `UploadBaseAction.upload()` does not support it, so you could either extend that action - or you'd need to call the `OCCore.importFileNamed()` method directly.

By default, `OCCore.importFileNamed()` (which `UploadBaseAction.upload()` also uses under the hood) does move the file from the provided location into its internal storage, so that deletion is not necessary if the `placeholderCompletionHandler` does not indicate an error.

Reply:
@felix-schwarz fixed

Reply:
@mneuwert I saw you extended `UploadBaseAction.upload()` to also call the completion block from the `OCCore.importFileNamed(placeholderCompletionHandler)`.

The regular `completionHandler` (that is called after the actual upload has finished) and the `placeholderCompletionHandler` (that is called after updating the local metadata table/item cache) should be strictly separate, however. Otherwise, the completion block passed to `UploadBaseAction.upload()` will be called more than once most of the time.

Reply:
@felix-schwarz fixed, added second call-back in  `UploadBaseAction.upload()`